Add Ott semantics tool grammar

k4rtik opened the pull request:
<!--- Briefly describe your changes in the field above. -->

## Description
<!--- If necessary, go into depth of what this pull request is doing. -->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- Feel free to remove whole sections, not points within the sections, that do not apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] **I am adding a new language.**
  - [x] The extension of the new language is used in hundreds of repositories on GitHub.com.
    - Search results for each extension:
      <!-- Replace FOOBAR with the new extension, and KEYWORDS with keywords unique to the language. Repeat for each extension added. -->
      -  https://github.com/search?utf8=%E2%9C%93&type=Code&ref=searchresults&q=extension%3Aott+grammar+NOT+nothack
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - https://github.com/ott-lang/ott/blob/master/tests/test10typing_minimal.ott
    - Sample license(s): https://github.com/ott-lang/ott/blob/master/LICENCE
  <!-- Update the Lightshow URLs below to show the grammar in action if you included one. -->
  - [x] I have included a syntax highlighting grammar: https://github-lightshow.herokuapp.com/
    - https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=from-url&grammar_format=auto&grammar_url=https%3A%2F%2Fgithub.com%2FJoeyEremondi%2Fvscode-ott%2Fblob%2Fmaster%2Fsyntaxes%2Fott.tmLanguage.json&grammar_text=&code_source=from-url&code_url=https%3A%2F%2Fgithub.com%2Fott-lang%2Fott%2Fblob%2Fmaster%2Ftests%2Ftest10typing_minimal.ott&code=
  - [ ] I have updated the heuristics to distinguish my language from others using the same extension.

lildude commented:
Usage doesn't meet our minimum requirements for merging yet, but it is close so I'll label as pending popularity:

Total files found: 808
Unique public user/repos: 180
Unique owners: 141

In the mean time, your `languages.yml` entry is missing a few fields too (I think the test will pick up a few of these). I also recommend adding a colour.

k4rtik commented:
Thanks for taking a look, @lildude, I wasn't sure what was ace mode needed for so I had skipped it. But I have added it now along with color.

About usage, ott is used by academics for writing papers, so oftentimes it lives in private repos (I just made two of my repos with ott files public). But I understand the concern and am happy to wait for a bit.

k4rtik commented:
Hi @lildude, the license for `vscode-ott` is the accepted 3-clause BSD license. Is there a way I can specify it so the CI does not complain about it thinking it's "other"?

Alhadis commented:
@k4rtik How did you add the grammar to this PR? Did you use the [`add-grammar`](https://github.com/github/linguist/blob/8b3c18c62a08ed90d8f5a3bb74966928fa915f80/script/add-grammar) script as advised in our [contributor's guide](https://github.com/github/linguist/blob/HEAD/CONTRIBUTING.md#adding-a-language), or did you register the submodule manually? Either way, adding support for a new language isn't something that should be done without first [setting up](https://github.com/github/linguist/blob/master/CONTRIBUTING.md#getting-started) a proper development environment. Doing this will also allow you to run tests locally (and solve any issues that CI would've flagged).

k4rtik commented:
I did use the add-grammar script the first time around following the contributing guide. I found the `vendor/licenses/git_submodule/vscode-ott.dep.yml` file lying around in my local branch yesterday. I am not sure how it was generated, I thought I should add it to this PR as there were similar files for other languages, so I added it yesterday and modified it today.

What is the command I should run to test locally? Do I just try and add grammar again after deleting the submodule?

Alhadis commented:
Alright, just checking. We sometimes get folks who attempt to do everything via the in-browser editor, so I needed to rule that out of the equation.

> What is the command I should run to test locally? Do I just try and add grammar again after deleting the submodule?

~~~sh
# To run all tests:
$ bundle exec rake test

# To run tests in a specific file:
$ bundle exec rake test TEST=test/test_pedantic.rb
~~~

> I found the `vendor/licenses/git_submodule/vscode-ott.dep.yml` file lying around in my local branch yesterday. I am not sure how it was generated, I thought I should add it to this PR as there were similar files for other languages, so I added it yesterday and modified it today.

🤔 That should have been included by [`script/add-grammar`](https://github.com/github/linguist/blob/0fca6f46491731d9adedc678d8663aa74d40a5a6/script/add-grammar#L134), unless there was an error during execution (which would've meant the call to `bundle exec licensed cache -c vendor/licenses/config.yml` was never reached). Do you remember if there was a problem when you first ran the script?

k4rtik commented:
Thanks, the tests are all passing locally.

```
Finished in 32.266520s, 46.1779 runs/s, 940.9444 assertions/s.

1490 runs, 30361 assertions, 0 failures, 0 errors, 0 skips
```

I did have some errors the first time around. IIRC, one of them was about `tm_scope: source.ott` missing in the Ott spec that I was adding in `languages.yml`. Most likely scenario is that the file was generated after the `add-grammar` script but I may have missed adding it to the commit. All good now, I guess.

(As an aside, I tend to not add things to a spec that I don't understand, hoping there will be safe defaults or complaints from some tests/CI, if some field was absolutely necessary. Having some explanation in CONTRIBUTING.md for required fields certainly would have helped.)

Alhadis commented:
> IIRC, one of them was about `tm_scope: source.ott` missing in the Ott spec that I was adding in `languages.yml`

That might be it. Yes, you're right, this could have been clarified in our docs. IMHO, the header of `languages.yml` ought to be revised so it's clear to users what fields are mandatory and which aren't. It doesn't help that descriptions aren't listed in alphabetical order, either.

> Most likely scenario is that the file was generated after the `add-grammar` script but I may have missed adding it to the commit

One thing the script (currently) doesn't do very well is clean up after itself in the event of a runtime error. This can leave one's branch in an unstable state, and complicate follow-up attempts to re-add the grammar.

k4rtik commented:
Hi @Alhadis, do you need any more changes from my end?

I know this PR will be pending popularity, but perhaps the tests workflow could proceed?

Alhadis commented:
@k4rtik I checked out your branch locally and ran the tests. No errors were encountered, indicating that CI will pass.

<details><summary><b>Output</b></summary>

~~~console
λ GitHub-Linguist (HEAD detached at fork/ott): brake samples test
cp lib/linguist/languages.yml tmp/x86_64-darwin20/stage/lib/linguist/languages.yml
cp lib/linguist/samples.json tmp/x86_64-darwin20/stage/lib/linguist/samples.json
install -c tmp/x86_64-darwin20/linguist/3.0.2/linguist.bundle lib/linguist/linguist.bundle
cp tmp/x86_64-darwin20/linguist/3.0.2/linguist.bundle tmp/x86_64-darwin20/stage/lib/linguist/linguist.bundle
Run options: --seed 4438

­# Running:

............................................................................................................
............................................................................................................
............................................................................................................
............................................................................................................
............................................................................................................
............................................................................................................
............................................................................................................
............................................................................................................
............................................................................................................
............................................................................................................
............................................................................................................
............................................................................................................
............................................................................................................
......................................................................................

Finished in 18.537892s, 80.3759 runs/s, 1637.7806 assertions/s.

1490 runs, 30361 assertions, 0 failures, 0 errors, 0 skips
~~~

</details>

Getting CI to run is kind of irrelevant at this point, since an extension or language needs adequate popularity before a PR to add support for it can be considered.
